src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -1169,34 +1340,35 @@ private boolean isParamOrVar(Symbol sym) {
      *  - A 'this' identifier qualified by a class name whose type is 'currentClass' or a supertype
      *    but also NOT an enclosing outer class of 'currentClass'.
      */
-    private boolean isExplicitThisReference(Types types, Type.ClassType currentClass, JCTree tree) {
-        switch (tree.getTag()) {
-            case PARENS:
-                return isExplicitThisReference(types, currentClass, TreeInfo.skipParens(tree));
-            case IDENT:
-            {
-                JCIdent ident = (JCIdent)tree;
-                Names names = ident.name.table.names;
-                return ident.name == names._this;
-            }
-            case SELECT:
-            {
-                JCFieldAccess select = (JCFieldAccess)tree;
-                Type selectedType = types.erasure(select.selected.type);
-                if (!selectedType.hasTag(CLASS))
-                    return false;
-                ClassSymbol currentClassSym = (ClassSymbol)((Type.ClassType)types.erasure(currentClass)).tsym;
-                ClassSymbol selectedClassSym = (ClassSymbol)((Type.ClassType)selectedType).tsym;
-                Names names = select.name.table.names;
-                return currentClassSym.isSubClass(selectedClassSym, types) &&
-                        (select.name == names._super ||
-                        (select.name == names._this &&
-                            (currentClassSym == selectedClassSym ||
-                            !currentClassSym.isEnclosedBy(selectedClassSym))));
-            }
-            default:
-                return false;
+    private boolean isExplicitThisReference(Types types, Type.ClassType currentClass, JCFieldAccess select) {
```

Comment:
suggestion: why not moving this to TreeInfo already in preparation for the changes coming with flexible constructor bodies?

Reply:
Good idea - thanks. Done in 304a92b619e.